### DISCUSSION TITLE
Ctimer and context

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#1717] [Q&A] [answered]
The [Ctimer docs](https://github.com/contiki-ng/contiki-ng/wiki/Documentation:-Timers#the-ctimer-library) call out:

> Do not assume any specific process context in the callback unless you are sure about how the callback timers are scheduled.

What does this mean? What can or can't our code do in the callback function?

**Comment by atiselsts** [comment, marked as answer]
This means that you can't assume that `PROCESS_CURRENT()` called from the callback's function is going to return the application's main function.

However, there is a workaround. The function `ctimer_set_with_process` is not documented in that wiki page for some reason. If you use this function to schedule the ctimer, the callback function of that timer is going to run in the context of the process that was passed as an argument to that function:
https://contiki-ng.readthedocs.io/en/master/_api/group__ctimer.html#gaee6c10c997cd9de6d1e82c211503521c
